geopandas/tools/sjoin.py

```diff
@@ -482,9 +545,32 @@ def _nearest_query(
             distances = np.array([], dtype=np.float64)
         else:
             distances = None
+
+    if on_attribute:
+        for attr in on_attribute:
+            (l_idx, r_idx), shared_attribute_rows = _filter_shared_attribute(
+                left_df, right_df, l_idx, r_idx, attr
+            )
+            distances = distances[shared_attribute_rows]
+
     return (l_idx, r_idx), distances
 
 
+def _filter_shared_attribute(left_df, right_df, l_idx, r_idx, attribute):
+    """
+    Returns the indices for the left and right dataframe that share the same entry
+    in the attribute column. Also returns a Boolean `shared_attribute_rows` for rows
+    with the same entry.
+    """
+    shared_attribute_rows = (
+        left_df.loc[l_idx][attribute].values == right_df.loc[r_idx][attribute].values
```

Comment:
```suggestion
        left_df.iloc[l_idx][attribute].values == right_df.iloc[r_idx][attribute].values
```

`l_idx` and `r_idx` are referring to the integer position, not the actual index.

Reply:
Thanks for picking up that mistake. Will keep that in mind